prometheus-dns-registration-service/main.go

```diff
@@ -0,0 +1,358 @@
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+	"github.com/pkg/errors"
+
+	log "github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// This structure is used to decode the configmap
+type config struct {
+	Global struct {
+		EvaluationInterval string `yaml:"evaluation_interval"`
+		ScrapeInterval     string `yaml:"scrape_interval"`
+		ScrapeTimeOut      string `yaml:"scrape_timeout"`
+	} `yaml:"global"`
+	RuleFiles     []string `yaml:"rule_files"`
+	ScrapeConfigs []struct {
+		HonorLabels bool   `yaml:"honor_labels"`
+		JobName     string `yaml:"job_name"`
+		MetricsPath string `yaml:"metrics_path"`
+		Perams      struct {
+			Match []string `yaml:"match[]"`
+		} `yaml:"params"`
+		ScrapeInterval string `yaml:"scrape_interval"`
+		StaticConfigs  []struct {
+			Targets []string `yaml:"targets"`
+		} `yaml:"static_configs"`
+	} `yaml:"scrape_configs"`
+}
+
+type clusterConfig struct {
+	ClusterName              string
+	MasterEndpoint           string
+	CertificateAuthorityData string
+	Session                  *session.Session
+}
+
+type clientConfig struct {
+	Client      *clientcmdapi.Config
+	ClusterName string
+	ContextName string
+	roleARN     string
+	sts         stsiface.STSAPI
+}
+
+func main() {
+	lambda.Start(handler)
+}
+
+func handler() {
+
+	hostedZoneID := os.Getenv("HOSTED_ZONE_ID")
+	clusterName := os.Getenv("CLUSTER_NAME")
+	configMapName := os.Getenv("CONFIG_MAP_NAME")
+
+	log.Info("Getting the k8s client")
+	clientSet, err := k8sClient(clusterName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Info("Successfully got the k8s client")
+
+	log.Info("Getting Prometheus server configmap")
+	configMap, err := clientSet.CoreV1().ConfigMaps("monitoring").Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		log.Fatalf("Unable to get the Prometheus server configmap, %v", err)
+	}
+	log.Info("Successfully got the Prometheus server configmap")
+
+	data := (configMap.Data["prometheus.yml"])
+	C := config{}
+	log.Info("Decoding configmap data into structure")
+	C, err = decodeConfig(data, C)
+	if err != nil {
+		log.Fatalf("Unable to decode into struct, %v", err)
+	}
+	log.Info("Successfully decoded configmap data into structure")
+
+	log.Info("Getting existing Route53 records")
+	targets, err := getRoute53Records(hostedZoneID)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	dataNew, err := updateTargets(C, targets)
+	if err != nil {
+		log.Fatal(err)
+	}
+	configMap.Data["prometheus.yml"] = string([]byte(dataNew))
+
+	_, err = clientSet.CoreV1().ConfigMaps("monitoring").Update(configMap)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Info("Successfully updated Prometheus server configmap")
+}
+
+// updateTargets is used to replace existing configmap Prometheus targets with new ones
+func updateTargets(C config, targets []string) ([]byte, error) {
+	log.Info("Replacing existing targets with updated values")
+	C.ScrapeConfigs[0].StaticConfigs[0].Targets = targets
```

Comment:
This line contains two direct references to a value in a slice. Are we sure that each slice has at least one value in it here? If not, Go will panic.

Reply:
Yeap each slice has at least one as this is a preconfigured map.